### PR TITLE
GCN for facebook dataset classification 

### PR DESCRIPTION
**Pull request by ananya-dub**
# Pull Request for semi-supervised node classification on facebook dataset using GCN 
**Name**: Ananya Dubey 
**Student Number**: 44913083 

## Overview 
This pull request is a working solution for semi supervised nodal classification for the Facebook dataset. The model implements a two-layer GCN using Pytorch Geometric's `GCNonv` with batch normalization. ADAM is used as the optimiser with a learning rate of 0.05 and cross entropy loss was used as the loss function. The model achieved an accuracy of 93.9% on the test set. 

## Usage: 
1. Download the `facebook.npz` file and place it within the `recognition\GCN_s4491308` folder. 
2. Update this file path in the `train.py` and `predict.py files` in the file_path variable before loading the dataset. 
3. Then run the `train.py` file and the `predict.py` file.

## Files Added and descriptions: 

- **README.md**: Overview of the project and the model implemented as well as the results. 
- **dataset.py**: File to load the facebook.npz data and split it into train/test/validation sets.
- **modules.py**: File to define the GCN model using pytorch-geometric.
- **train.py**: File to train the model and plot the validation and training losses. 
- **predict.py**: File to run inference on the test set and plot the tSNE visualisation of the node embeddings generated by the model.
- **tSNE-embedding.png**: Plot for the tSNE embeddings 
- **training_validation_loss.png**: Plot for the training and validation losses 
- **GCN_image.jpg**: Descriptive image for the algorithm 

Please let me know if there is any feedback or anything that needs to be changed for the merge. 

Thanks, 
Ananya

### CODE REVIEW

**Comment by shaivikaaaa** [comment]
@gayanku

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
GNN FB Page-Page Network dataset ---- Normal Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 5 | |
| Implementation functions as intended	| 3 | 2.5 |  2 layer GCN, torch_geometric/GCNConv |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 0 | Normal Difficulty |
| Section IV : Max mark 15 from 20		|   | **14.5**   | |

**TSNE / UMAP:** 
- Good.
- Discussion: Good, with a bit more depth useful.

**Suggestions:**
Hard coded hyper parameters should be avoided (epochs)
Code can have some early stop, ex: train loop should save best model by keeping track of current best accuracy.
Could add a hyper parameter search or discuss how the values were obtained.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. No early stop.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. Some incorrect usage instructions.</td><td>-1</td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-7</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by ananya-dub** [comment]
Hi Gayan, 

I had an approved extension for the pull request submission. Would I still loose the extra 2 marks? 

Thanks, 
Ananya

**Comment by shakes76** [comment]
Approved extension +2
